ixmp4/data/db/optimization/variable/repository.py

```diff
@@ -0,0 +1,185 @@
+        variable_id : int
+            The id of the :class:`ixmp4.data.abstract.optimization.Variable`.
+        column_name : str
+            The name of the Column, which must be unique in connection with the names of
+            :class:`ixmp4.data.abstract.Run` and
+            :class:`ixmp4.data.abstract.optimization.Variable`.
+        indexset_name : str
+            The name of the :class:`ixmp4.data.abstract.optimization.IndexSet` the
+            Column will be linked to.
+        \*\*kwargs: any
+            Keyword arguments to be passed to
+            :func:`ixmp4.data.abstract.optimization.Column.create`.
+        """
+        indexset = self.backend.optimization.indexsets.get(
+            run_id=run_id, name=indexset_name
+        )
+        self.columns.create(
+            name=column_name,
+            constrained_to_indexset=indexset.id,
+            dtype=pd.Series(indexset.elements).dtype.name,
+            variable_id=variable_id,
+            unique=True,
+            **kwargs,
+        )
+
+    def add(
+        self,
+        run_id: int,
+        name: str,
+    ) -> Variable:
+        variable = Variable(name=name, run__id=run_id)
+        variable.set_creation_info(auth_context=self.backend.auth_context)
+        self.session.add(variable)
+
+        return variable
+
+    @guard("view")
+    def get(self, run_id: int, name: str) -> Variable:
+        exc = db.select(Variable).where(
+            (Variable.name == name) & (Variable.run__id == run_id)
+        )
+        try:
+            return self.session.execute(exc).scalar_one()
+        except db.NoResultFound:
+            raise Variable.NotFound
+
+    @guard("view")
+    def get_by_id(self, id: int) -> Variable:
+        obj = self.session.get(self.model_class, id)
+
+        if obj is None:
+            raise Variable.NotFound(id=id)
+
+        return obj
+
+    @guard("edit")
+    def create(
+        self,
+        run_id: int,
+        name: str,
+        constrained_to_indexsets: str | list[str] | None = None,
+        column_names: list[str] | None = None,
+        **kwargs,
+    ) -> Variable:
+        # Convert to list to avoid enumerate() splitting strings to letters
+        if isinstance(constrained_to_indexsets, str):
+            constrained_to_indexsets = list(constrained_to_indexsets)
+        if column_names:
+            # TODO If this is removed, need to check above that constrained_to_indexsets
+            #  is not None
+            if constrained_to_indexsets is None:
+                raise ValueError(
+                    "Received `column_names` to name columns, but no "
+                    "`constrained_to_indexsets` to indicate which IndexSets to use for "
+                    "these columns. Please provide `constrained_to_indexsets` or "
+                    "remove `column_names`!"
+                )
+            elif len(column_names) != len(constrained_to_indexsets):
+                raise ValueError(
```

Comment:
I'm kind of confused as to how this error is reraised on the frontend, i think this should be a custom ixmp4 error class which is automatically serialized. Not sure why this is working with `ValueError`, i dont think it should. Maybe warrants an investigation.

Reply:
To be honest, I don't see why this shouldn't work with `ValueError`. Could you elaborate, please?